src/linear_algebra/bilinear_form.lean

```diff
@@ -1016,3 +1034,328 @@ begin
+  have : n ∈ submodule.span K ({x} : set V) ⊔
+    B.orthogonal (submodule.span K ({x} : set V)) :=
+    (span_sup_ortho_eq_top hB₂ hx).symm ▸ submodule.mem_top,
+  rcases submodule.mem_sup.1 this with ⟨y, hy, z, hz, rfl⟩,
+  specialize hm ⟨z, hz⟩,
+  rw restrict_def at hm,
+  erw [add_right, show B m.1 y = 0, by exact m.2 y hy, hm, add_zero]
+end
+
+/- Let V be a finite dimensional vector space over the field K with the
+  nondegenerate bilinear form B. Then for all m ∈ M, f_m : M → R₁ : n ↦ B m n is
+  a linear functional in the dual space.
+
+  Furthermore, the map, φ : M → M* : m ↦ f_m is an isomorphism. -/
+
+/-- Given a bilinear form `B` on the modual `M`, `to_dual' B` is the linear map
+  from `M` to its dual such that `to_dual B m` is the functional `λ x, B m x`. -/
+def to_dual' (B : bilin_form R₃ M₃) : M₃ →ₗ[R₃] module.dual R₃ M₃ :=
+  B.to_lin
+
+@[simp] lemma to_dual'_def {B : bilin_form R₃ M₃} {m n : M₃} :
+  B.to_dual' m n = B m n := rfl
+
+lemma to_dual'_injective {B : bilin_form R₃ M₃} (hB : B.nondegenerate) :
+  function.injective (to_dual' B) :=
+to_lin_injective hB
+
+open finite_dimensional
+
+variable [finite_dimensional K V]
+
+-- In order to show that `to_dual` is a surjective map we used the fact that
+-- the dimensions of a vector space equal to the dimensions of its dual.
+-- So rather than working with modules over rings, we work with vector spaces
+lemma to_dual'_bijective (B : bilin_form K V) (hB : B.nondegenerate) :
+  function.bijective (to_dual' B) :=
+begin
+  classical,
+  refine ⟨to_dual'_injective hB, _⟩,
+  change function.surjective B.to_dual',
+  refine (linear_map.injective_iff_surjective_of_findim_eq_findim
+    (linear_equiv.findim_eq _)).1 (to_dual'_injective hB),
+  have hB := classical.some_spec (exists_is_basis_finite K V),
+  haveI := classical.choice hB.2,
+  exact is_basis.to_dual_equiv _ hB.1
+end
+
+/-- To dual is the `linear_equiv` with the underlying linear map `to_dual'`. -/
+noncomputable def to_dual (B : bilin_form K V) (hB : B.nondegenerate) :
+  V ≃ₗ[K] module.dual K V :=
+{ map_smul' := B.to_dual'.map_smul',
+  .. add_equiv.of_bijective B.to_dual'.to_add_monoid_hom (to_dual'_bijective B hB) }
+
+lemma to_dual_def {B : bilin_form K V} (hB : B.nondegenerate) {m n : V} :
+  B.to_dual hB m n = B m n := rfl
+
+-- We start proving that symmetric nondegenerat bilinear forms are diagonalisable, or equivilently
+-- there exists a orthogonal basis with respect to any symmetric nondegenerate bilinear form.
+
+lemma findim_ortho_span_singleton
+  {B : bilin_form K V} (hB : sym_bilin_form.is_sym B)
+  {x : V} (hx : ¬ B.is_ortho x x) : findim K V =
+    findim K (B.orthogonal (submodule.span K ({x} : set V))) + 1 :=
+begin
+  rw [← submodule.findim_quotient_add_findim (submodule.span K ({x} : set V)),
+      findim_span_singleton
+        (show x ≠ 0, by exact λ hx', hx (hx'.symm ▸ zero_left _)),
+      (quotient_equiv_of_ortho_singleton hB hx).findim_eq]
```

Comment:
I think a separate `have` makes sense here
```suggestion
  have : x ≠ 0 := λ hx', hx (hx'.symm ▸ zero_left _)),
  rw [← submodule.findim_quotient_add_findim (submodule.span K ({x} : set V)),
      findim_span_singleton this, (quotient_equiv_of_ortho_singleton hB hx).findim_eq]
```
perhaps you still need the `by exact`, I didn't check,